Fix bug where a function that is a property of a nested type throws

ketchel opened the pull request:
Right now a state with the following format will throw an error:
```
const obj = {
    'name': 'outer',
    'outerFunc': () => console.log('outer func'),
    'outerSymbol': Symbol('outer'),
    'inner': {
        'name': 'inner',
        'innerFunc': () => console.log('inner func'),
        'innerSymbol': Symbol('inner'),
    }
}
```
Because the `innerFunc` and `innerSymbol` aren't remove before attempting the `postMessage`. Callling `JSON.stringify` and then `JSON.parse` on the result yields an object without the unsupported types and solves this issue.

ketchel committed:
Use JSON to remove unsupported types, add partialize and rehydrate

Romainlg29 commented:
Hi @ketchel , 
I don't think using JSON.stringigy and parse, are the most performant way to do that?
If the store has a lot like thousands of objects, it would take some time.

What do you think about implementing a deep?: number | boolean = false; properties in the options, that allows nested properties?
If a number is provided, it checks recursively until it reaches the number.
If true is provided, maybe JSON.stringify / parse?
And false, the default value, would not check the nested objects?

ketchel commented:
Hi @Romainlg29, thanks for the reply.
You raise a good point, so I decided to compare the current and JSON approaches as well as a new recursive approach. I also added a flag to the recursive function that, when `false`, won't check nested objects, similar to what you suggested. I created a jsfiddle if you want to look at the code and test these as well: https://jsfiddle.net/sbhazgwp

Here were my findings:
![image](https://github.com/user-attachments/assets/f7f8e95e-560e-4bb4-af1d-b0d8fb4a32d6)

I think the current approach is being severely impacted by the spread operator. When I switched it to a `delete` it became much faster. Let me know which approach you prefer and I'll do some more testing before updating this fork. Personally, I would go with the JSON approach as it is the most robust despite being slightly slower than the recursive approach. I still need to test whether mutating the object returned from `get()` will cause problems (though I suspect it won't) and whether having array items replace with `null` or `empty` will cause issues. The recursive approach turns non serializable array items into `empty` items and the JSON approach replaces them with `null`.

Romainlg29 commented:
Hi @ketchel ,

Thank you for testing. The JSON method is pretty good!
I just have some concerns, if a function or symbols are stored in array, it would replace them with null or undefined? So we cannot guarantee the type safety here?

```ts
JSON.stringify(["hello", () => {}, "World"]);
["hello", null, "World"];
```

So if the user has something like myarraywithafunction.at(1), this would be null and break things?

If you have an idea about this, tell me! 
Otherwise, I'm ok with your method 🙂

Romainlg29 commented:
Maybe we can just write a warning in the readme or something like that when the user want to use this feature ?

ketchel commented:
Yes I think that's a good option. We can't delete the non-serializable items from the array either because then `myarraywithafunction.at(1)` will return the wrong item or throw an index out of bounds exception. The other thing we could do is allow the user to pass in a [replacer](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#replacer) and a [reviver](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#replacer) which would give them absolute control over how things are serialized if they so desire.

I also think we should get rid of the spread operator on line 169 now that we know it's a bit slow. What do you think?

Romainlg29 commented:
Yes, you're right.
I think that the best way is to let the user make their own functions (replacer/reviver).
Yes, we can remove the spread operator!

Do you want to implement the changes?

ketchel commented:
I would like to but it might take me a a couple of weeks to get around to it if that's alright

Romainlg29 commented:
Ok, I'll do it next week 👍

ketchel commented:
Hey @Romainlg29! I had some spare time today so I implemented the replacer and the reviver. The replacer fan be a function or array of keys that should be kept , so I used that to eliminate the spread operator by keeping essentially the same code you had before but using it to collect a list of the keys that have changed.

There's still an issue surrounding changes to properties in nested objects. On the current commit, nested objects will be clobbered in other tabs when they change in the current one. I think we have four options:
1. Don't sync nested objects
2. Always send objects in the postmessage
3. Implement a recursive function to go deep in the state object to get the list of keys with modified values.
4. Don't check which keys are changed at all. Always send the entire state object in the `postmessage`

I think number 3 and 4 are the best options. Personally, I'd go with 4. I suspect the performance impact will be minimal and if someone has a huge object that is causing performance issues, they could always pass in a `replacer` that would eliminate it from the `postmessage`. I'd be happy to compare the performance of these approaches, but I wanted to ask what you thought first.

You might also notice I modified your example, that was just for ease of testing and I can revert it before we merge my fork, but I thought you might want to try it out so I kept it in this commit.

Thanks :)

Romainlg29 commented:
Hi @ketchel ,
I'm ok with number 4. Could you still check if the performance is ok when doing that?

I was wondering if we could take inspiration from the persist middleware and let the user completely handle these cases using `partialize` / `rehydrate`? We could provide two functions where the TS definitions would be: `(state: StoreType) => Partial<StoreType>`.
This approach would be a bit simpler, and would make users responsible for handling the edge cases.

What do you think?

ketchel commented:
Sure thing, I'll lyk what I find.
Just to make sure I understand, you're suggesting we allow users to pass a `partialize` method that gets called on the state before the `postmessage`, and `rehydrate` that gets called on the state received by a tab? If so, I like that idea. My only concern is `partialize` and `replacer` would do more or less the same thing. We could:

1. Rename `replacer` to `partialize` and explain the difference between this version and zustand's version in the documentation, the difference being that `replacer` takes key value pairs instead of an object.
2. Remove the replacer and reviver arguments entirely and use the more zustand like approach.
3. Include both partialize and replacer.

1 is fast but potentially confusing for users, 2 is probably slow, and 3 is flexible but redundant. I'm torn on this one, what do you think?

Edit: Performance without checking for changed keys is actually better, go figure

Romainlg29 committed:
✨ feat: partialize / rehydrate

Romainlg29 commented:
Hi @ketchel ,

I have made a little branch to expose my idea, you can find it here: #20 
It would provide great control, and keep legacy support for the previous versions.
Tell me what's your trought about that.

Sorry for responding late.

ketchel committed:
Remove rehydrate, use replacer instead of spread operator

ketchel committed:
Merge example working, update readme, refactor out sendCurrentStateToOtherTabs

ketchel committed:
Whitespace

ketchel committed:
Unnecessary exclamation point

ketchel committed:
Revert example

ketchel commented:
Hey @Romainlg29!

Sorry for the wait. I like your idea a lot! Simple and everyone using this library is probably familiar with persist already. I rebased onto your branch and got an example working. I also changed `rehydrate` to `merge` because that made more sense to me, lmk what you think. I don't think keeping the `Object.entries.reduce` approach is neccessary for legacy support, the current state of the fork will behave in the same way as previous versions, but if you want to include it I think we should `replacer` instead of the spread operator for performance reasons.

The only sticking point is that `JSON.stringify` will throw an error if the user has a `BigInt` in their store (or if there's a circular reference in their object but I'm not as worried about that). You can get around this by converting the `BigInt` to a string and then back to a `BigInt` after parsing, so we could do that for the first level the state, but I think it's a pretty niche use case and users can do that using `partialize` and `merge` so I don't think it's necessary. But, if there is someone using the library right now who has a `BigInt` in their state and they update their version of the package it would stop working, which isn't ideal.

I also updated the README. Lmk what you think about `merge` instead of `rehydrate` and handling the `BigInt` case, thanks.

Edit: I implemented a potential solution to continue supporting BigInt. Now it will break if one of the strings in the user's state ends with `'use-broadcast-ts-bigint-\uFFFF'` where `\uFFFF` is a unicode non-character but I don't think that would be an issue lol. I'll leave the decision about whether to keep it with you :)

ketchel committed:
Continue to support BitInts, Refactor dup code into sendChangeToOtherTabs

ketchel committed:
Update readme

Romainlg29 commented:
Hi @ketchel ,
I have tested your changes, and I'm ready to merge your changes!
Everything I tested works fine.

For the BigInt issue, I think that we can let the developer handle the case. Despite this, I think that we should specify that in the readme?
As I really don't want to break existing applications, we're going to release it as a 2.0.0. Are you Ok with that?

Moreover, maybe we can create a small contribution section at the end of the readme 🙂 ?

Let me know,
Romain.

fspoettel commented:
Hi @ketchel / @Romainlg29, thank you both for your work on this PR and this library. It is exactly what I needed for [an application](https://arkham.build/) I'm building, so I went ahead and added a vendored version of this PR's code to the application to test it.

The one feedback I would raise is that with a medium to large state, the `state = JSON.parse(JSON.stringify(state));` operation adds a meaningful amount of work on the main thread, even after partializing it. In my case, this ranged in the 20-30ms range, 3x-4x the amount of time that the BroadcastChannel itself needed to process the message.

It would be nice if this could be optional / on the user, given that it is not needed in all cases.

ketchel committed:
Add option to skip JSON.parse(JSON.stringify)

ketchel committed:
Bump version

ketchel committed:
Remove console.log

ketchel commented:
Hey @Romainlg29, that's great! I agree about BigInts. I added it to the readme. Releasing as v2.0.0 seems like a good idea. I'm not sure what to write in a contributing section, but I made you a collaborator on my fork so that if you want to write something there. I think you'll have to be the one to merge since I don't have write access to the main repo :).

@fspoettel good idea, I added an option to skip serialization.

Romainlg29 commented:
@ketchel @fspoettel It's now available under use-broadcast-ts@2.0.0 🙂
Thank you so much @ketchel for your work!

ketchel commented:
Thank you @Romainlg29! I appreciate your patience and collaborative spirit throughout this process. All the best :)
